### PR TITLE
#163900813 add register candidate endpoint 

### PR DESCRIPTION
**Pull request by Muart-C**
**What does this PR do?**
 - Add a register a candidate endpoint.

**Description of Task to be completed?**
- Add functionality to check if the user exists before registration.
- Add functionality to check if a party exists before registering a candidate.
- Add functionality to check if an office exists before registering a candidate.

**How should the app be tested?**
-  `git clone https://github.com/Muart-C/politico.git`

- `cd politico`

- `git checkout ft-register-candidate-163900813`

-  `$ export FLASK_APP="run.py"`
   `$ export FLASK_ENV="development"`
   `$ export FLASK_DEBUG=1`
   `$ export DATABASE_POLITICO="politico"`
   `$ export DATABASE_USERNAME="yourpostgresusername"`
   `$ export DATABASE_PASSWORD="yourpostgrespassword"`
   `$ export DATABASE_HOST="localhost"`
   `$ export DATABASE_PORT=5432`
 - Or alternatively you can create a .env file and add the above configurations on your local machine.

-  Then start your app via the command `flask run` 

- First of all, ensure you install Postman - Test client to test the API endpoint.

[Run in Postman](https://app.getpostman.com/run-collection/4e6c88cc481fc0ceefd0) via your web browser.

**Steps to follow on postman**

- Send POST requests to create a user, a party and an office to create a party.
- A candidate should be registered and a 201 response status returned.
 else a 404 error will be returned if either the party, office and the user was not registered

**What are the relevant pivotal tracker stories**
https://www.pivotaltracker.com/story/show/163900813

**screenshot**
screenshot from 2019-02-19 10-57-29

### CODE REVIEW

File: api/database/database.py [hunk 1]
```diff
@@ -14,109 +19,66 @@ def __init__(self):
...
-    #     self.cursor.execute("DELETE FROM users;")
-    #     self.cursor.execute("DELETE FROM parties;")
-    #     self.cursor.execute("DELETE FROM candidates;")
-    #     self.cursor.execute("DELETE FROM petitions;")
-    #     self.cursor.execute("DELETE FROM votes;")
+    def fetch_a_single_data_row(self, query):
+        """retrieve a single data row."""
+        print(self.cursor.execute(query))
+        row = self.cursor.fetchone()
+        # self.cursor.close()
```
Comment (meshack-mbuvi) [inline api/database/database.py]: Why is this code commented out?

File: api/database/db_manage.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    return [delete_user_data, delete_parties_data,delete_offices_data,\
+         delete_candidates_data, delete_candidates_data,\
+              delete_petitions_data, delete_votes_data]
+
+
+def drop_tables_if_exists():
+    # drop existing tables if there are any
+    drop_tables = '''DROP TABLE IF EXISTS users,parties,offices,candidates,petitions,votes CASCADE;'''
+    return [drop_tables]
```
Comment (meshack-mbuvi) [inline api/database/db_manage.py]: Add a new line at the end of **every** file

File: api/models/candidates_model.py [hunk 3]
```diff
@@ -10,9 +11,15 @@ def __init__(self, office_id, party_id, candidate_id):
         self.candidate_id = candidate_id
 
     def create_a_candidate(self):
+        """check if an office exists first."""
+        # office = Office(name=None,office_type=None).get_office(self.office_id)
+        # print(office)
```
Comment (meshack-mbuvi) [inline api/models/candidates_model.py]: Remove code that you have commented out; it's not necessary

File: api/models/offices_model.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 """#!api/models/offices_model.py"""
```
Comment (meshack-mbuvi) [inline api/models/offices_model.py]: The comment on **line 1** is not necessary

File: api/models/users_model.py [hunk 5]
```diff
@@ -1,33 +1,33 @@
...
+        self.othername = kwargs.get('othername')
+        self.passport_url = kwargs.get('passport_url')
+        self.phone_number =  kwargs.get('phone_number')
+        self.hash_password =  kwargs.get('password')
 
     def create_user(self):
         """create a user if one does not exist."""
         self.cursor.execute('''SELECT * FROM users WHERE email='{}';'''.format(self.email))
         user=self.cursor.fetchone()
+        print(user)
```
Comment (meshack-mbuvi) [inline api/models/users_model.py]: Remove the print statement too.

File: api/database/database.py (hunk 1 above)
Comment (Muart-C) [reply api/database/database.py]: yeah, I will uncomment it thanks for the observation.

**Pull request merged**